File: 837. New 21 Game/new21Game.java
class Solution {
    public double new21Game(int N, int K, int W) {
        double[] dp = new double[K + W];
        double sum = 0;
        for (int i = K; i < K + W; i++) {
            dp[i] = (i <= N) ? 1.0 : 0.0;
            sum += dp[i];
        }
        for (int i = K - 1; i > -1; i--) {
            dp[i] = sum / W;
            sum += dp[i] - dp[i + W];
        }
        return dp[0];
    }
}

File: 837. New 21 Game/new21Game.py
class Solution:
    def new21Game(self, N: int, K: int, W: int) -> float:
        dp = [0] * (K + W)
        s = 0
        for i in range(K, K + W):
            if i <= N:
                dp[i] = 1
            else:
                dp[i] = 0
            s += dp[i]
        
        for i in range(K - 1, -1, -1):
            dp[i] = s / W
            s = dp[i] + s - dp[i + W]
        return dp[0]